src/core/SampleBuffer.cpp

```diff
@@ -1017,30 +1017,57 @@ void SampleBuffer::visualize(
 	const float ySpace = h * 0.5f;
 	const int nbFrames = focusOnRange ? toFrame - fromFrame : m_frames;
 
-	const int fpp = qBound<int>(1, nbFrames / w, 20);
-	QPointF * l = new QPointF[nbFrames / fpp + 1];
-	QPointF * r = new QPointF[nbFrames / fpp + 1];
-	int n = 0;
+	const int fpp = qMax<int>(1, nbFrames / w);
+	const int totalPoints = nbFrames / fpp * 2 + 2;
+	std::vector<QPointF> fMax(totalPoints);
+	std::vector<QPointF> fRms(totalPoints);
+	int curFrame = 0;
 	const int xb = dr.x();
 	const int first = focusOnRange ? fromFrame : 0;
 	const int last = focusOnRange ? toFrame - 1 : m_frames - 1;
 
 	for (int frame = first; frame <= last; frame += fpp)
 	{
+		float maxData = -1;
+		float minData = 1;
+
+		float rmsData[2] = {0, 0};
+
+		// Find maximum and minimum samples within range
+		for (int i = 0; i < fpp && frame + i < last; ++i)
+		{
+			for (int j = 0; j < 2; ++j)
+			{
+				auto curData = m_data[frame + i][j];
+
+				if (curData > maxData) { maxData = m_data[frame + i][j]; }
+				if (curData < minData) { minData = m_data[frame + i][j]; }
+
+				rmsData[j] += curData * curData;
+			}
+		}
+
+		const float trueRmsData = (rmsData[0] + rmsData[1]) * 0.5 / fpp;
+		const float sqrtRmsData = sqrt(trueRmsData);
+		const float maxRmsData = qBound(minData, sqrtRmsData, maxData);
+		const float minRmsData = qBound(minData, -sqrtRmsData, maxData);
+
 		auto x = xb + ((frame - first) * double(w) / nbFrames);
 		// Partial Y calculation
 		auto py = ySpace * m_amplification;
-		l[n] = QPointF(x, (yb - (m_data[frame][0] * py)));
-		r[n] = QPointF(x, (yb - (m_data[frame][1] * py)));
-		++n;
+		fMax[curFrame*2] = QPointF(x, (yb - (maxData * py)));
+		fMax[curFrame*2+1] = QPointF(x, (yb - (minData * py)));
+		fRms[curFrame*2] = QPointF(x, (yb - (maxRmsData * py)));
+		fRms[curFrame*2+1] = QPointF(x, (yb - (minRmsData * py)));
+		++curFrame;
 	}
 
-	p.setRenderHint(QPainter::Antialiasing);
-	p.drawPolyline(l, nbFrames / fpp);
-	p.drawPolyline(r, nbFrames / fpp);
+	//p.setRenderHint(QPainter::Antialiasing);
+	p.drawPolyline(fMax.data(), totalPoints);
```

Comment:
Is the extra `+ 2` in `totalPoints` going to change from what it was before (`nbFrames / fpp * 2`)?